src/core/i2c.zig

```diff
@@ -0,0 +1,121 @@
+const std = @import("std");
+const micro = @import("microzig.zig");
+const chip = @import("chip");
+
+pub fn I2CController(comptime index: usize) type {
+    const SystemI2CController = chip.I2CController(index);
+
+    const I2CDevice = struct {
+        const Device = @This();
+
+        internal: SystemI2CController,
+        address: u7,
+
+        const Direction = enum { read, write };
+        fn Transfer(comptime direction: Direction) type {
+            return switch (direction) {
+                .read => struct {
+                    const Self = @This();
+
+                    device: *const Device,
+
+                    pub const Reader = std.io.Reader(Self, ReadError, readSome);
+
+                    pub fn reader(self: Self) Reader {
+                        return Reader{ .context = self };
```

Comment:
With this implementation, several reads will perform restarted reads instead of a single continuous read. These are semantically different in I²C

Reply:
'Fixed' in commit [009d3fc](https://github.com/ZigEmbeddedGroup/microzig/pull/26/commits/009d3fc3fa5c7636197f04cadad7448ffb845b70) and later: The second read call on an STM32F303 will now panic, since it is basically impossible to implement a multi-part read.

Note that a multi-part write will work on STM32F303, at least if the number of bytes is < 256; I cannot really test >= 256 writes (single-part or multi-part), but I'm not at all sure they will work...

Both have been achieved by doing a 'lazy' (re)START during the first read or write call-- that really seems to be the only way the current API can be properly implemented on STM32 microcontrollers...